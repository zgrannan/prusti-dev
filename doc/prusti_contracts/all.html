<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../prusti_contracts/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../prusti_contracts/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location">Crate prusti_contracts</h2></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../prusti_contracts/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><h1 class="fqn"><span class="in-band">List of all items</span></h1><h3 id="Structs">Structs</h3><ul class="structs docblock"><li><a href="struct.Ghost.html">Ghost</a></li><li><a href="struct.Int.html">Int</a></li><li><a href="struct.Map.html">Map</a></li><li><a href="struct.Seq.html">Seq</a></li></ul><h3 id="Macros">Macros</h3><ul class="macros docblock"><li><a href="macro.body_invariant.html">body_invariant</a></li><li><a href="macro.closure.html">closure</a></li><li><a href="macro.ghost.html">ghost</a></li><li><a href="macro.map.html">map</a></li><li><a href="macro.predicate.html">predicate</a></li><li><a href="macro.prusti_assert.html">prusti_assert</a></li><li><a href="macro.prusti_assume.html">prusti_assume</a></li><li><a href="macro.seq.html">seq</a></li></ul><h3 id="Attribute-Macros">Attribute Macros</h3><ul class="attributes docblock"><li><a href="attr.after_expiry.html">after_expiry</a></li><li><a href="attr.assert_on_expiry.html">assert_on_expiry</a></li><li><a href="attr.ensures.html">ensures</a></li><li><a href="attr.extern_spec.html">extern_spec</a></li><li><a href="attr.ghost_constraint.html">ghost_constraint</a></li><li><a href="attr.invariant.html">invariant</a></li><li><a href="attr.model.html">model</a></li><li><a href="attr.pure.html">pure</a></li><li><a href="attr.refine_trait_spec.html">refine_trait_spec</a></li><li><a href="attr.requires.html">requires</a></li><li><a href="attr.trusted.html">trusted</a></li></ul><h3 id="Functions">Functions</h3><ul class="functions docblock"><li><a href="fn.before_expiry.html">before_expiry</a></li><li><a href="fn.exists.html">exists</a></li><li><a href="fn.forall.html">forall</a></li><li><a href="fn.old.html">old</a></li><li><a href="fn.prusti_set_union_active_field.html">prusti_set_union_active_field</a></li></ul></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="prusti_contracts" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0-nightly (f9cba6374 2022-07-31)" ></div></body></html>