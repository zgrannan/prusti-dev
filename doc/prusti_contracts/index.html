<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `prusti_contracts` crate."><meta name="keywords" content="rust, rustlang, rust-lang, prusti_contracts"><title>prusti_contracts - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../crates.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../prusti_contracts/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../prusti_contracts/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Crate prusti_contracts</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.1.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#functions">Functions</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../prusti_contracts/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">prusti_contracts</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/prusti_contracts/lib.rs.html#1-354">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.body_invariant.html" title="prusti_contracts::body_invariant macro">body_invariant</a></div><div class="item-right docblock-short"><p>A macro for writing a loop body invariant.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.closure.html" title="prusti_contracts::closure macro">closure</a></div><div class="item-right docblock-short"><p>A macro for defining a closure with a specification.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ghost.html" title="prusti_contracts::ghost macro">ghost</a></div><div class="item-right docblock-short"><p>A macro for defining ghost blocks which will be left in for verification
but omitted during compilation.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.map.html" title="prusti_contracts::map macro">map</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.predicate.html" title="prusti_contracts::predicate macro">predicate</a></div><div class="item-right docblock-short"><p>A macro for defining a predicate using prusti expression syntax instead
of just Rust expressions.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.prusti_assert.html" title="prusti_contracts::prusti_assert macro">prusti_assert</a></div><div class="item-right docblock-short"><p>A macro for writing assertions using the full prusti specifications</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.prusti_assume.html" title="prusti_contracts::prusti_assume macro">prusti_assume</a></div><div class="item-right docblock-short"><p>A macro for writing assumptions using prusti syntax</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.seq.html" title="prusti_contracts::seq macro">seq</a></div><div class="item-right docblock-short"></div></div></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Ghost.html" title="prusti_contracts::Ghost struct">Ghost</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Int.html" title="prusti_contracts::Int struct">Int</a></div><div class="item-right docblock-short"><p>a mathematical (unbounded) integer type
it should not be constructed from running rust code, hence the private unit inside</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Map.html" title="prusti_contracts::Map struct">Map</a></div><div class="item-right docblock-short"><p>A map type</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Seq.html" title="prusti_contracts::Seq struct">Seq</a></div><div class="item-right docblock-short"><p>A sequence type</p>
</div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.before_expiry.html" title="prusti_contracts::before_expiry fn">before_expiry</a></div><div class="item-right docblock-short"><p>This function is used to evaluate an expression in the context just
before the borrows expires.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.exists.html" title="prusti_contracts::exists fn">exists</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.forall.html" title="prusti_contracts::forall fn">forall</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.old.html" title="prusti_contracts::old fn">old</a></div><div class="item-right docblock-short"><p>This function is used to evaluate an expression in the “old”
context, that is at the beginning of the method call.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.prusti_set_union_active_field.html" title="prusti_contracts::prusti_set_union_active_field fn">prusti_set_union_active_field</a></div><div class="item-right docblock-short"></div></div></div><h2 id="attributes" class="small-section-header"><a href="#attributes">Attribute Macros</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.after_expiry.html" title="prusti_contracts::after_expiry attr">after_expiry</a></div><div class="item-right docblock-short"><p>A macro for writing a pledge on a function.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.assert_on_expiry.html" title="prusti_contracts::assert_on_expiry attr">assert_on_expiry</a></div><div class="item-right docblock-short"><p>A macro for writing a two-state pledge on a function.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ensures.html" title="prusti_contracts::ensures attr">ensures</a></div><div class="item-right docblock-short"><p>A macro for writing a postcondition on a function.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.extern_spec.html" title="prusti_contracts::extern_spec attr">extern_spec</a></div><div class="item-right docblock-short"><p>A macro for specifying external functions.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ghost_constraint.html" title="prusti_contracts::ghost_constraint attr">ghost_constraint</a></div><div class="item-right docblock-short"><p>A macro to add trait bounds on a generic type parameter and specifications
which are active only when these bounds are satisfied for a call.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.invariant.html" title="prusti_contracts::invariant attr">invariant</a></div><div class="item-right docblock-short"><p>A macro for type invariants.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.model.html" title="prusti_contracts::model attr">model</a></div><div class="item-right docblock-short"><p>Macro for creating type models.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.pure.html" title="prusti_contracts::pure attr">pure</a></div><div class="item-right docblock-short"><p>A macro for marking a function as pure.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.refine_trait_spec.html" title="prusti_contracts::refine_trait_spec attr">refine_trait_spec</a></div><div class="item-right docblock-short"><p>A macro for impl blocks that refine trait specifications.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.requires.html" title="prusti_contracts::requires attr">requires</a></div><div class="item-right docblock-short"><p>A macro for writing a precondition on a function.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.trusted.html" title="prusti_contracts::trusted attr">trusted</a></div><div class="item-right docblock-short"><p>A macro for marking a function as trusted.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="prusti_contracts" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0-nightly (f9cba6374 2022-07-31)" ></div></body></html>