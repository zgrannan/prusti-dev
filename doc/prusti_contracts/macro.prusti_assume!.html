<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.prusti_assume.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.prusti_assume.html">macro.prusti_assume.html</a>...</p>
    <script>location.replace("macro.prusti_assume.html" + location.search + location.hash);</script>
</body>
</html>