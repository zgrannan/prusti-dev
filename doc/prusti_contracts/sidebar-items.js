window.SIDEBAR_ITEMS = {"attr":[["after_expiry","A macro for writing a pledge on a function."],["assert_on_expiry","A macro for writing a two-state pledge on a function."],["ensures","A macro for writing a postcondition on a function."],["extern_spec","A macro for specifying external functions."],["ghost_constraint","A macro to add trait bounds on a generic type parameter and specifications which are active only when these bounds are satisfied for a call."],["invariant","A macro for type invariants."],["model","Macro for creating type models."],["pure","A macro for marking a function as pure."],["refine_trait_spec","A macro for impl blocks that refine trait specifications."],["requires","A macro for writing a precondition on a function."],["trusted","A macro for marking a function as trusted."]],"fn":[["before_expiry","This function is used to evaluate an expression in the context just before the borrows expires."],["exists",""],["forall",""],["old","This function is used to evaluate an expression in the “old” context, that is at the beginning of the method call."],["prusti_set_union_active_field",""]],"macro":[["body_invariant","A macro for writing a loop body invariant."],["closure","A macro for defining a closure with a specification."],["ghost","A macro for defining ghost blocks which will be left in for verification but omitted during compilation."],["map",""],["predicate","A macro for defining a predicate using prusti expression syntax instead of just Rust expressions."],["prusti_assert","A macro for writing assertions using the full prusti specifications"],["prusti_assume","A macro for writing assumptions using prusti syntax"],["seq",""]],"struct":[["Ghost",""],["Int","a mathematical (unbounded) integer type it should not be constructed from running rust code, hence the private unit inside"],["Map","A map type"],["Seq","A sequence type"]]};