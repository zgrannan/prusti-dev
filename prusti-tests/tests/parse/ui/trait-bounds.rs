// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Phide_uuids=true -Pno_verify=true
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// normalize-stdout-test: "\[[a-z0-9]{4}\]::" -> "[$(CRATE_ID)]::"

extern crate prusti_contracts;
use prusti_contracts::*;

struct Foo<'a, T: PartialEq, const L: usize>(&'a [T; L]);

impl<'a, T: PartialEq, const L: usize> Foo<'a, T, L> {
    pub fn bar(self) -> &'a [T; L] { self.0 }
}

#[extern_spec]
impl<'a, T: PartialEq, const L: usize> Foo<'a, T, L> {
    #[pure]
    #[ensures(result == self.0)]
    pub fn bar(self) -> &'a [T; L];
}

fn main() {}
