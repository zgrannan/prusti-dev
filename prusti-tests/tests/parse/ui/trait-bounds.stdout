// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Phide_uuids=true -Pno_verify=true
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// normalize-stdout-test: "/[[a-z0-9]{4}/]::" -> "[$(CRATE_ID)]::"





#![feature(register_tool)]
#![register_tool(prusti)]
#[prelude_import]
use std::prelude::rust_2018::*;
#[macro_use]
extern crate std;
extern crate prusti_contracts;
use prusti_contracts::*;
struct Foo<'a, T: PartialEq, const L : usize>(&'a [T; L]);
impl <'a, T: PartialEq, const L : usize> Foo<'a, T, L> {
    pub fn bar(self) -> &'a [T; L] { self.0 }
}
#[allow(non_camel_case_types)]
struct PrustiStructFoo_$(NUM_UUID)<'a, T: PartialEq,
                                                        const L :
                                                        usize>(&'a ::core::marker::PhantomData<()>,
                                                               ::core::marker::PhantomData<T>);
impl <'a, T: PartialEq, const L : usize>
 PrustiStructFoo_$(NUM_UUID)<'a, T, L> {
    #[allow(unused_must_use, unused_variables, dead_code)]
    #[prusti::spec_only]
    #[prusti::spec_id = "$(NUM_UUID)"]
    #[prusti::assertion =
      "{/"kind/":{/"Expr/":{/"spec_id/":/"$(UUID)/",/"expr_id/":101}}}"]
    fn prusti_post_item_bar_$(NUM_UUID)(_self:
                                                                 Foo<'a, T,
                                                                     L>,
                                                             result:
                                                                 &'a [T; L]) {

        #[prusti::spec_only]
        #[prusti::expr_id = "$(NUM_UUID)_101"]
        || -> bool { result == _self.0 };
    }
    #[prusti::pure]
    #[prusti::post_spec_id_ref = "$(NUM_UUID)"]
    #[prusti::trusted]
    #[prusti::extern_spec]
    pub fn bar(_self: Foo<'a, T, L>) -> &'a [T; L] {
        Foo::<'a, T, L>::bar(_self);
        ::core::panicking::panic("not implemented")
    }
}
fn main() {}
Procedure(ProcedureSpecification { pres: [], posts: [Assertion { kind: Expr(Expression { spec_id: SpecificationId($(UUID)), id: ExpressionId(101), expr: DefId(0:32 ~ trait_bounds[$(CRATE_ID)]::{impl#1}::prusti_post_item_bar_$(NUM_UUID)::{closure#0}) }) }], pledges: [], predicate_body: None, pure: true, trusted: true })
